File: modules/editor/web/js/ballerina/ast/binary-expression.js
/**
 * Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 *
 * WSO2 Inc. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
define(['lodash', './expression'], function (_, Expression) {

    /**
     * Constructor for BinaryExpression
     * @param {Object} args - Arguments to create the BinaryExpression
     * @constructor
     * @augments Expression
     */
    var BinaryExpression = function (args) {
        Expression.call(this, 'BinaryExpression');
        this._operator = _.get(args, 'operator');
    };

    BinaryExpression.prototype = Object.create(Expression.prototype);
    BinaryExpression.prototype.constructor = BinaryExpression;

    /**
     * setting parameters from json
     * @param {Object} jsonNode to initialize from
     */
    BinaryExpression.prototype.initFromJson = function (jsonNode) {
        this.setExpression(this.generateExpressionString(jsonNode));
    };

    /**
     * Generates the binary expression as a string.
     * @param {Object} jsonNode - A node explaining the structure of binary expression.
     * @return {string} - Arguments as a string.
     * @private
     */
    BinaryExpression.prototype.generateExpressionString = function (jsonNode) {
        var self = this;
        var expString = "";

        for (var itr = 0; itr < jsonNode.children.length; itr++) {
            var childJsonNode = jsonNode.children[itr];
            //TODO : Need to remove this if/else ladder by delegating expression string calculation to child classes
            if (childJsonNode.type == "basic_literal_expression") {
                if(childJsonNode.basic_literal_type == "string") {
                    // Adding double quotes if it is a string.
                    expString += "\"" + childJsonNode.basic_literal_value + "\"";
                } else {
                    expString += childJsonNode.basic_literal_value;
                }
            } else if (childJsonNode.type == "variable_reference_expression") {
                expString += childJsonNode.variable_reference_name;
            } else {
                var child = self.getFactory().createFromJson(childJsonNode);
                child.initFromJson(childJsonNode);
                expString += child.getExpression();
            }

            if (itr !== jsonNode.children.length - 1) {
                expString += " " + this.getOperator() + " ";
            }
        }
        return expString;
    };

    BinaryExpression.prototype.getOperator = function (){
        return this._operator;
    };

    return BinaryExpression;
});

File: modules/ballerina-core/src/main/java/org/wso2/ballerina/core/model/expressions/BinaryExpression.java
/*
*  Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
*
*  WSO2 Inc. licenses this file to you under the Apache License,
*  Version 2.0 (the "License"); you may not use this file except
*  in compliance with the License.
*  You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
*  Unless required by applicable law or agreed to in writing,
*  software distributed under the License is distributed on an
*  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
*  KIND, either express or implied.  See the License for the
*  specific language governing permissions and limitations
*  under the License.
*/
package org.wso2.ballerina.core.model.expressions;

import org.wso2.ballerina.core.model.NodeExecutor;
import org.wso2.ballerina.core.model.Operator;
import org.wso2.ballerina.core.model.Position;
import org.wso2.ballerina.core.model.values.BValue;
import org.wso2.ballerina.core.model.values.BValueType;

import java.util.function.BiFunction;

/**
 * {@code BinaryExpression} represents a binary expression.
 *
 * @see BinaryCompareExpression
 * @see BinaryArithmeticExpression
 * @see BinaryLogicalExpression
 * @see BinaryEqualityExpression
 * @since 0.8.0
 */
public class BinaryExpression extends UnaryExpression {

    protected Expression lExpr;
    protected BiFunction<BValueType, BValueType, BValueType> evalFuncNewNew;

    public BinaryExpression(Expression lExpr, Operator op, Expression rExpr, Position location) {
        super(op, rExpr, location);
        this.lExpr = lExpr;
    }

    public Expression getLExpr() {
        return lExpr;
    }

    public BiFunction<BValueType, BValueType, BValueType> getEvalFunc() {
        return evalFuncNewNew;
    }

    public void setEvalFunc(BiFunction<BValueType, BValueType, BValueType> evalFuncNewNew) {
        this.evalFuncNewNew = evalFuncNewNew;
    }

    public BValue execute(NodeExecutor executor) {
        return executor.visit(this);
    }

    public void setRExpr(Expression rExpr) {
        this.rExpr = rExpr;
    }

}
